## Data Wrangling - Identify missing data - part 1

### What is the purpose of data wrangling?

Data wrangling is the process of converting data from the initial format to a format that may be better for analysis.

### What is the fuel consumption (L/100k) rate for the diesel car?

**"Automobile Dataset"** link: https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data.

In [25]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt

In [13]:
# create headers 
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

In [16]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'

In [11]:
# before adding headers
df.head()

3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.60  ...   
0  3    ?  alfa-romero  gas  std   two  convertible  rwd  front   88.6  ...  \
1  1    ?  alfa-romero  gas  std   two    hatchback  rwd  front   94.5  ...   
2  2  164         audi  gas  std  four        sedan  fwd  front   99.8  ...   
3  2  164         audi  gas  std  four        sedan  4wd  front   99.4  ...   
4  2    ?         audi  gas  std   two        sedan  fwd  front   99.8  ...   

   130  mpfi  3.47  2.68  9.00  111  5000  21  27  13495  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
1  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
2  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
3  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  
4  136  mpfi  3.19  3.40   8.5  110  5500  19  25  15250  

[5 rows x 26 columns]

In [21]:
# after adding headers
df = pd.read_csv(url, names =headers)
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors   
0          3                 ?  alfa-romero       gas        std          two  \
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size   
0  convertible          rwd           front        88.6  ...          130  \
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg   
0         mpfi  3.47    2.68               9.0        111      5000       21  \
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

As we can see, several question marks appeared in the dataframe; those are missing values which may hinder our further analysis.

<div>So, how do we identify all those missing values and deal with them?</div> 

<b>How to work with missing data?</b>

Steps for working with missing data:

<ol>
    <li>Identify missing data</li>
    <li>Deal with missing data</li>
    <li>Correct data format</li>
</ol>


<h2 id="identify_handle_missing_values">Identify and handle missing values</h2>

<h3 id="identify_missing_values">Identify missing values</h3>
<h4>Convert "?" to NaN</h4>
In the car dataset, missing data comes with the question mark "?".
We replace "?" with NaN (Not a Number), Python's default missing value marker for reasons of computational speed and convenience. Here we use the function: 
 <pre>.replace(A, B, inplace = True) </pre>
to replace A by B.


In [26]:
# replace "?" to NaN
df.replace("?", np.nan, inplace = True)
df.head(5)

symboling normalized-losses         make fuel-type aspiration num-of-doors   
0          3               NaN  alfa-romero       gas        std          two  \
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size   
0  convertible          rwd           front        88.6  ...          130  \
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg   
0         mpfi  3.47    2.68               9.0        111      5000       21  \
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

<h4>Evaluating for Missing Data</h4>

The missing values are converted by default. We use the following functions to identify these missing values. There are two methods to detect missing data:

<ol>
    <li><b>.isnull()</b></li>
    <li><b>.notnull()</b></li>
</ol>
The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data.

In [38]:
missing_data = df.isnull()
missing_data.head()

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors   
0      False               True  False      False       False         False  \
1      False               True  False      False       False         False   
2      False               True  False      False       False         False   
3      False              False  False      False       False         False   
4      False              False  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size   
0       False         False            False       False  ...        False  \
1       False         False            False       False  ...        False   
2       False         False            False       False  ...        False   
3       False         False            False       False  ...        False   
4       False         False            False       False  ...        False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm   
0        False  False   False              False       False     False  \
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  

[5 rows x 26 columns]

"True" means the value is a missing value while "False" means the value is not a missing value.


<h4>Count missing values in each column</h4>
<p>
Using a for loop in Python, we can quickly figure out the number of missing values in each column. As mentioned above, "True" represents a missing value and "False" means the value is present in the dataset.  In the body of the for loop the method ".value_counts()" counts the number of "True" values. 
</p>


In [43]:
# for one column
df['symboling'].value_counts()

symboling
 0    67
 1    54
 2    32
 3    27
-1    22
-2     3
Name: count, dtype: int64

In [52]:
df.shape

(205, 26)

In [56]:
# for all use for loop

for column in missing_data.columns.values.tolist():
#     print(len(column))
    print(missing_data[column].value_counts())
    print(" ")

symboling
False    205
Name: count, dtype: int64
 
normalized-losses
False    164
True      41
Name: count, dtype: int64
 
make
False    205
Name: count, dtype: int64
 
fuel-type
False    205
Name: count, dtype: int64
 
aspiration
False    205
Name: count, dtype: int64
 
num-of-doors
False    203
True       2
Name: count, dtype: int64
 
body-style
False    205
Name: count, dtype: int64
 
drive-wheels
False    205
Name: count, dtype: int64
 
engine-location
False    205
Name: count, dtype: int64
 
wheel-base
False    205
Name: count, dtype: int64
 
length
False    205
Name: count, dtype: int64
 
width
False    205
Name: count, dtype: int64
 
height
False    205
Name: count, dtype: int64
 
curb-weight
False    205
Name: count, dtype: int64
 
engine-type
False    205
Name: count, dtype: int64
 
num-of-cylinders
False    205
Name: count, dtype: int64
 
engine-size
False    205
Name: count, dtype: int64
 
fuel-system
False    205
Name: count, dtype: int64
 
bore
False    201
True       4
Na

Based on the summary above, each column has 205 rows of data and seven of the columns containing missing data:

<ol>
    <li>"normalized-losses": 41 missing data</li>
    <li>"num-of-doors": 2 missing data</li>
    <li>"bore": 4 missing data</li>
    <li>"stroke" : 4 missing data</li>
    <li>"horsepower": 2 missing data</li>
    <li>"peak-rpm": 2 missing data</li>
    <li>"price": 4 missing data</li>
</ol>


<h3 id="deal_missing_values">Deal with missing data</h3>
<b>How to deal with missing data?</b>

<ol>
    <li>Drop data<br>
        a. Drop the whole row<br>
        b. Drop the whole column
    </li>
    <li>Replace data<br>
        a. Replace it by mean<br>
        b. Replace it by frequency<br>
        c. Replace it based on other functions
    </li>
</ol>

Whole columns should be dropped only if most entries in the column are empty. In our dataset, none of the columns are empty enough to drop entirely.
We have some freedom in choosing which method to replace data; however, some methods may seem more reasonable than others. We will apply each method to many different columns:

<b>Replace by mean:</b>

<ul>
    <li>"normalized-losses": 41 missing data, replace them with mean</li>
    <li>"stroke": 4 missing data, replace them with mean</li>
    <li>"bore": 4 missing data, replace them with mean</li>
    <li>"horsepower": 2 missing data, replace them with mean</li>
    <li>"peak-rpm": 2 missing data, replace them with mean</li>
</ul>

<b>Replace by frequency:</b>

<ul>
    <li>"num-of-doors": 2 missing data, replace them with "four". 
        <ul>
            <li>Reason: 84% sedans is four doors. Since four doors is most frequent, it is most likely to occur</li>
        </ul>
    </li>
</ul>

<b>Drop the whole row:</b>

<ul>
    <li>"price": 4 missing data, simply delete the whole row
        <ul>
            <li>Reason: price is what we want to predict. Any data entry without price data cannot be used for prediction; therefore any row now without price data is not useful to us</li>
        </ul>
    </li>
</ul>


<h4>Calculate the mean value for the "normalized-losses" column </h4>

In [60]:
avg_norm_loss = df['normalized-losses'].astype('float').mean(axis=0)
avg_norm_loss

122.0

<h4>Replace "NaN" with mean value in "normalized-losses" column</h4>

In [63]:
df['normalized-losses'].replace(np.nan,avg_norm_loss , inplace=True)

In [65]:
# change in dataframe
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors   
0          3             122.0  alfa-romero       gas        std          two  \
1          3             122.0  alfa-romero       gas        std          two   
2          1             122.0  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size   
0  convertible          rwd           front        88.6  ...          130  \
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg   
0         mpfi  3.47    2.68               9.0        111      5000       21  \
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

<h4>Calculate the mean value for the "bore" column</h4>


In [67]:
avg_bore = df['bore'].astype('float').mean(axis=0)
avg_bore

3.3297512437810943

In [68]:
df['bore'].replace(np.nan, avg_bore, inplace=True)

In [69]:
df(sample)

symboling normalized-losses         make fuel-type aspiration   
0            3             122.0  alfa-romero       gas        std  \
1            3             122.0  alfa-romero       gas        std   
2            1             122.0  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...   
0            two  convertible          rwd           front        88.6  ...  \
1            two  convertible          rwd           front        88.6  ...   
2            two    hatchback          rwd           front        94.5  ...   
3           four        sedan          fwd           front        99.8  ...   
4           four        sedan          4wd           front        99.4  ...   
..           ...          ...          ...             ...         ...  ...   
200         four        sedan          rwd           front       109.1  ...   
201         four        sedan          rwd           front       109.1  ...   
202         four        sedan          rwd           front       109.1  ...   
203         four        sedan          rwd           front       109.1  ...   
204         four        sedan          rwd           front       109.1  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower   
0            130         mpfi  3.47    2.68               9.0        111  \
1            130         mpfi  3.47    2.68               9.0        111   
2            152         mpfi  2.68    3.47               9.0        154   
3            109         mpfi  3.19    3.40              10.0        102   
4            136         mpfi  3.19    3.40               8.0        115   
..           ...          ...   ...     ...               ...        ...   
200          141         mpfi  3.78    3.15               9.5        114   
201          141         mpfi  3.78    3.15               8.7        160   
202          173         mpfi  3.58    2.87               8.8        134   
203          145          idi  3.01    3.40              23.0        106   
204          141         mpfi  3.78    3.15               9.5        114   

     peak-rpm city-mpg highway-mpg  price  
0        5000       21          27  13495  
1        5000       21          27  16500  
2        5000       19          26  16500  
3        5500       24          30  13950  
4        5500       18          22  17450  
..        ...      ...         ...    ...  
200      5400       23          28  16845  
201      5300       19          25  19045  
202      5500       18          23  21485  
203      4800       26          27  22470  
204      5400       19          25  22625  

[205 rows x 26 columns]

In [74]:
df['bore'].isnull().sum()

0

In [75]:
# before replace
df['stroke'].isnull().sum() #output 4

4

In [76]:
avg_stroke = df['stroke'].astype('float').mean(axis=0)
avg_stroke

3.255422885572139

In [78]:
df['stroke'].replace(np.nan, avg_stroke , inplace=True)

In [80]:
# after replace values
df['stroke'].isnull().sum()

0

In [81]:
df['horsepower'].isnull().sum()  #output 2

2

In [84]:
avg_horsepower = df['horsepower'].astype('float').mean(axis=0)
avg_horsepower

104.25615763546799

In [86]:
df['horsepower'].replace(np.nan, avg_horsepower , inplace=True )

In [87]:
df['horsepower'].isnull().sum()

0

In [88]:
df['peak-rpm'].isnull().sum() #before ,, output 2

2

In [89]:
peakrpm_avg = df['peak-rpm'].astype('float').mean(axis=0)
peakrpm_avg

5125.369458128079

In [91]:
df['peak-rpm'].replace(np.nan, peakrpm_avg, inplace=True)

In [92]:
df['peak-rpm'].isnull().sum() #after replacing ,, output 0 

0

In [93]:
# for all use for loop

for column in missing_data.columns.values.tolist():
#     print(len(column))
    print(missing_data[column].value_counts())
    print(" ")

symboling
False    205
Name: count, dtype: int64
 
normalized-losses
False    164
True      41
Name: count, dtype: int64
 
make
False    205
Name: count, dtype: int64
 
fuel-type
False    205
Name: count, dtype: int64
 
aspiration
False    205
Name: count, dtype: int64
 
num-of-doors
False    203
True       2
Name: count, dtype: int64
 
body-style
False    205
Name: count, dtype: int64
 
drive-wheels
False    205
Name: count, dtype: int64
 
engine-location
False    205
Name: count, dtype: int64
 
wheel-base
False    205
Name: count, dtype: int64
 
length
False    205
Name: count, dtype: int64
 
width
False    205
Name: count, dtype: int64
 
height
False    205
Name: count, dtype: int64
 
curb-weight
False    205
Name: count, dtype: int64
 
engine-type
False    205
Name: count, dtype: int64
 
num-of-cylinders
False    205
Name: count, dtype: int64
 
engine-size
False    205
Name: count, dtype: int64
 
fuel-system
False    205
Name: count, dtype: int64
 
bore
False    201
True       4
Na

In [97]:
# check isnull values
df.isnull().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         2
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                4
dtype: int64

As we replaced all values by mean. 

#### Now 
<b>Replace by frequency:</b>

<ul>
    <li>"num-of-doors": 2 missing data, replace them with "four". 
        <ul>
            <li>Reason: 84% sedans is four doors. Since four doors is most frequent, it is most likely to occur</li>
        </ul>
    </li>
</ul>

<b>Drop the whole row:</b>

<ul>
    <li>"price": 4 missing data, simply delete the whole row
        <ul>
            <li>Reason: price is what we want to predict. Any data entry without price data cannot be used for prediction; therefore any row now without price data is not useful to us</li>
        </ul>
    </li>
</ul>


To see which values are present in a particular column, we can use the ".value_counts()" method:


In [98]:
df['num-of-doors'].value_counts()

num-of-doors
four    114
two      89
Name: count, dtype: int64

We can see that four doors are the most common type. We can also use the ".idxmax()" method to calculate the most common type automatically:


In [99]:
df['num-of-doors'].value_counts().idxmax()

'four'

The replacement procedure is very similar to what we have seen previously:


In [100]:
#replace the missing 'num-of-doors' values by the most frequent 
df["num-of-doors"].replace(np.nan, "four", inplace=True)

Finally, let's drop all rows that do not have price data:

In [101]:
# simply drop whole row with NaN in "price" column
df.dropna(subset=['price'], axis=0 , inplace=True)

In [102]:
df['price'].isnull().sum()

0

In [104]:
# reset index, because we droped two rows
df.reset_index(drop=True, inplace=True)

### Dataset With No Missing Values :') 

In [106]:
df.sample(10)

symboling normalized-losses         make fuel-type aspiration   
190         -2               103        volvo       gas        std  \
22           1               118        dodge       gas      turbo   
192         -2               103        volvo       gas        std   
2            1             122.0  alfa-romero       gas        std   
78           3               153   mitsubishi       gas        std   
34           0               110        honda       gas        std   
172         -1                65       toyota       gas        std   
148          1                74       toyota       gas        std   
31           1               101        honda       gas        std   
37           0               106        honda       gas        std   

    num-of-doors body-style drive-wheels engine-location  wheel-base  ...   
190         four      sedan          rwd           front       104.3  ...  \
22           two  hatchback          fwd           front        93.7  ...   
192         four      sedan          rwd           front       104.3  ...   
2            two  hatchback          rwd           front        94.5  ...   
78           two  hatchback          fwd           front        96.3  ...   
34          four      sedan          fwd           front        96.5  ...   
172         four      sedan          fwd           front       102.4  ...   
148         four  hatchback          fwd           front        95.7  ...   
31           two  hatchback          fwd           front        93.7  ...   
37           two  hatchback          fwd           front        96.5  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower   
190          141         mpfi  3.78    3.15               9.5        114  \
22            98         mpfi  3.03    3.39               7.6        102   
192          141         mpfi  3.78    3.15               9.5        114   
2            152         mpfi  2.68    3.47               9.0        154   
78           122         2bbl  3.35    3.46               8.5         88   
34            92         1bbl  2.91    3.41               9.2         76   
172          122         mpfi  3.31    3.54               8.7         92   
148           92         2bbl  3.05    3.03               9.0         62   
31            79         1bbl  2.91    3.07              10.1         60   
37           110         1bbl  3.15    3.58               9.0         86   

     peak-rpm city-mpg highway-mpg  price  
190      5400       23          28  12940  
22       5500       24          30   7957  
192      5400       24          28  15985  
2        5000       19          26  16500  
78       5000       25          32   8499  
34       6000       30          34   7295  
172      4200       27          32  10898  
148      4800       31          38   6488  
31       5500       38          42   5399  
37       5800       27          33   9095  

[10 rows x 26 columns]